exonum-java-binding/tutorials/car-registry/car-registry-service/src/main/java/com/example/car/MyService.java

```diff
@@ -0,0 +1,148 @@
+import com.exonum.binding.core.runtime.ServiceInstanceSpec;
+import com.exonum.binding.core.service.AbstractService;
+import com.exonum.binding.core.service.Configuration;
+import com.exonum.binding.core.service.Node;
+import com.exonum.binding.core.storage.database.Prefixed;
+import com.exonum.binding.core.storage.indices.ProofMapIndexProxy;
+import com.exonum.binding.core.transaction.ExecutionException;
+import com.exonum.binding.core.transaction.Transaction;
+import com.exonum.binding.core.transaction.TransactionContext;
+import com.google.inject.Inject;
+import io.vertx.ext.web.Router;
+import java.util.List;
+import java.util.Optional;
+
+public final class MyService extends AbstractService {
+
+  public static final int ADD_VEHICLE_TX_ID = 0;
+  public static final int CHANGE_OWNER_TX_ID = 1;
+  public static final byte ID_ALREADY_EXISTS_ERROR_CODE = 100;
+  public static final byte NO_VEHICLE_ERROR_CODE = 101;
+
+  @Inject
+  public MyService(ServiceInstanceSpec instanceSpec) {
+    super(instanceSpec);
+  }
+
+  // ci-block ci-createPublicApiHandlers {
+  @Override
+  public void createPublicApiHandlers(Node node, Router router) {
+    var apiController = new ApiController(this, node);
+    apiController.mount(router);
+  }
+  // }
+
+  // ci-block ci-add-vehicle {
+  @Transaction(ADD_VEHICLE_TX_ID)
+  public void addVehicle(Transactions.AddVehicle args, TransactionContext context) {
+    var serviceData = context.getServiceData();
+    addVehicle(args, serviceData);
+  }
+
+  private void addVehicle(AddVehicle args, Prefixed serviceData) {
+    var schema = new MySchema(serviceData);
+    ProofMapIndexProxy<String, Vehicle> vehicles = schema.vehicles();
+
+    // Check there is no vehicle with such id in the registry yet
+    var newVehicle = args.getNewVehicle();
+    var id = newVehicle.getId();
+    if (vehicles.containsKey(id)) {
+      var existingVehicle = vehicles.get(id);
+      var errorDescription = String.format("The registry already contains a vehicle "
+          + "with id (%s): existing=%s, new=%s", id, existingVehicle, newVehicle);
+      throw new ExecutionException(ID_ALREADY_EXISTS_ERROR_CODE, errorDescription);
+    }
+
+    // Add the vehicle to the registry
+    vehicles.put(id, newVehicle);
+  }
+  // }
+
+  // ci-block ci-change-owner {
+  @Transaction(CHANGE_OWNER_TX_ID)
+  public void changeOwner(Transactions.ChangeOwner args, TransactionContext context) {
+    var serviceData = context.getServiceData();
+    var schema = new MySchema(serviceData);
+    ProofMapIndexProxy<String, Vehicle> vehicles = schema.vehicles();
+
+    // Check the vehicle with such ID exists
+    var id = args.getId();
+    if (!vehicles.containsKey(id)) {
+      throw new ExecutionException(NO_VEHICLE_ERROR_CODE, "No vehicle with such id: " + id);
+    }
+
+    // Update the owner
+    // Get the current entry
+    var vehicleEntry = vehicles.get(id);
+    // Update the owner
+    var newOwner = args.getNewOwner();
```

Comment:
It will be good to add a check that this operation is performed by the car owner. And it will add some clarity about why do we need keys. But it will add complexity.

Reply:
Yes, we currently don't have a registry of car owners :-) We say this topic (authorization) is out of scope in the beginning, but it would be nice to cover in a more advanced tutorial/how to.